[BranchFolding] Fix missing predecessors of landing-pad

HaohaiWen opened the pull request:
When removing an empty machine basic block, all of its successors should
be inherited by its fall through MBB. This keeps CFG as only have one
entry which is required by LiveDebugValues.

Reland #77441 as LiveDebugValues test.

HaohaiWen committed:
[SEH][CodeGen] Add test to track CFG optimization bug for SEH (#77441)

LiveDebugValues requires CFG only has one entry. BranchFolding and
MachineBlockPlacement may remove all predecessors of landing pad which
leaves it to be another entry.

HaohaiWen committed:
[BranchFolding][SEH] Add test to track SEH CFG optimization

This test tracks BranchFolding pass which removes fall through jump and
leaves landing-pad to be machine basic block of no predecessors. It
would raise bug as introduced in #77441.

HaohaiWen committed:
[BranchFolding] Fix missing predecessors of landing-pad

When removing an empty machine basic block, all of its successors should
be inherited by its fall through MBB. This keeps CFG as only have one
entry which is required by LiveDebugValues.

phoebewang commented on llvm/lib/CodeGen/BranchFolding.cpp:
```diff
@@ -1363,6 +1363,14 @@ bool BranchFolder::OptimizeBlock(MachineBasicBlock *MBB) {
         MachineBasicBlock *Pred = *(MBB->pred_end()-1);
         Pred->ReplaceUsesOfBlockWith(MBB, &*FallThrough);
       }
+      // Add rest successors of MBB to successors of FallThrough. Those
+      // successors are not directly reachable via MBB, so it should be
+      // landing-pad.
+      for (auto SI = MBB->succ_begin(), SE = MBB->succ_end(); SI != SE; ++SI)
+        if (*SI != &*FallThrough && !FallThrough->isSuccessor(*SI)) {
+          assert((*SI)->isEHPad() && "Bad CFG");
+          FallThrough->copySuccessor(MBB, SI);
+        }
```
Can be put the code to `ReplaceUsesOfBlockWith`?

HaohaiWen replied:
We need to run these code before removing MBB. ReplaceUsesOfBlockWith just replace one of its users. It may run many needed to be run multi times.

phoebewang replied:
It's checked by `isSuccessor`, so multi call is not a problem.
One problem, no matter add here or in `ReplaceUsesOfBlockWith`, is whether these successors be removed latter? I think adding an edge to a died BB may also problematic.

HaohaiWen replied:
Let's see what happened before/after this code:
Before:
```
bb.0:
  successors: %bb.3(0x7ffff800), %bb.5(0x00000800); %bb.3(100.00%), %bb.5(0.00%)

  JMP_1 %bb.3

bb.2:
  successors: %bb.3(0x7ffff800), %bb.4(0x00000800); %bb.3(100.00%), %bb.4(0.00%)


bb.3:
; predecessors: %bb.2, %bb.0
  successors: %bb.6(0x80000000); %bb.6(100.00%)

  JMP_1 %bb.6

bb.4 (machine-block-address-taken, landing-pad, ehfunclet-entry):
; predecessors: %bb.2
  successors: %bb.5(0x80000000); %bb.5(100.00%)

  CLEANUPRET

bb.5 (landing-pad, ehfunclet-entry):
; predecessors: %bb.0, %bb.4

  CLEANUPRET

bb.6:
; predecessors: %bb.3

  RET 0

# End machine code for function main.
```
After
```
bb.0:
  successors: %bb.3(0x7ffff800), %bb.5(0x00000800); %bb.3(100.00%), %bb.5(0.00%)

  JMP_1 %bb.3

bb.2:
  successors: %bb.3(0x7ffff800), %bb.4(0x00000800); %bb.3(100.00%), %bb.4(0.00%)


bb.3:
; predecessors: %bb.2, %bb.0
  successors: %bb.6(0x80000000), %bb.4(0x00000800); %bb.6(100.00%), %bb.4(0.00%)

  JMP_1 %bb.6

bb.4 (machine-block-address-taken, landing-pad, ehfunclet-entry):
; predecessors: %bb.2, %bb.3
  successors: %bb.5(0x80000000); %bb.5(100.00%)

  CLEANUPRET

bb.5 (landing-pad, ehfunclet-entry):
; predecessors: %bb.0, %bb.4

  CLEANUPRET

bb.6:
; predecessors: %bb.3

  RET 0

# End machine code for function main.
```
As we can see, No BB is removed at this time. We're adding edge from one live BB to another live BB.
In fact, dead BB was removed by BranchFolder::RemoveDeadBlock later. The CHECK is the final result after many optimization.

phoebewang replied:
This doesn't answer the questions. My questions are:

- Why can't put the code in `ReplaceUsesOfBlockWith`? The `isSuccessor` can make sure the same BB won't add again;
- Is it possible the added BB might be removed latter. We don't have a mechanism to remove the dead successors. Would it be a problem if we keep an edge to dead BBs?

HaohaiWen replied:
> Why can't put the code in ReplaceUsesOfBlockWith? The isSuccessor can make sure the same BB won't add again;

It's possible MBB don't have any predecessor. e.g. The function entry BB is invoke llvm.seh.scope.begin. In such case, we still need to add bb.4 as successor of bb.3 before removing MBB.
```
bb.2:
  successors: %bb.3(0x7ffff800), %bb.4(0x00000800); %bb.3(100.00%), %bb.4(0.00%)


bb.3:
; predecessors: %bb.2
  successors: %bb.6(0x80000000); %bb.6(100.00%)

  JMP_1 %bb.6

bb.4 (machine-block-address-taken, landing-pad, ehfunclet-entry):
; predecessors: %bb.2
  successors: %bb.5(0x80000000); %bb.5(100.00%)

  CLEANUPRET

bb.5 (landing-pad, ehfunclet-entry):
; predecessors: %bb.4

  CLEANUPRET

bb.6:
; predecessors: %bb.3

  RET 0

# End machine code for function main.
```

> Is it possible the added BB might be removed latter. We don't have a mechanism to remove the dead successors. Would it be a problem if we keep an edge to dead BBs?

No, addSuccessor(), removeSuccessor() will properly maintain predecessors. We don't need to care about that.

phoebewang commented:
> This keeps CFG as only have one entry which is required by LiveDebugValues

If we just want to make LiveDebugValues happy, should it better to remove these died BB instead?

HaohaiWen committed:
Merge remote-tracking branch 'origin/main' into bf-fix

HaohaiWen commented:
> > This keeps CFG as only have one entry which is required by LiveDebugValues
> 
> If we just want to make LiveDebugValues happy, should it better to remove these died BB instead?

This died BB was already removed by BranchFolder. See rebased test diff. We just want to salvage its successors info.

phoebewang commented:
> > > This keeps CFG as only have one entry which is required by LiveDebugValues
> > 
> > 
> > If we just want to make LiveDebugValues happy, should it better to remove these died BB instead?
> 
> This died BB was already removed by BranchFolder. See rebased test diff. We just want to salvage its successors info.

I see. I thought the successors are removed too.

HaohaiWen committed:
Reland #77441

HaohaiWen commented on llvm/lib/CodeGen/BranchFolding.cpp:
```diff
@@ -1624,6 +1632,15 @@ bool BranchFolder::OptimizeBlock(MachineBasicBlock *MBB) {
+                   ++SI)
+                if (*SI != CurTBB && !CurTBB->isSuccessor(*SI)) {
+                  assert((*SI)->isEHPad() && "Bad CFG");
+                  CurTBB->copySuccessor(MBB, SI);
```
Before this code:
```
bb.0:
  successors: %bb.6(0x7ffff800), %bb.5(0x00000800); %bb.6(100.00%), %bb.5(0.00%)

  JMP_1 %bb.6

bb.3:
  successors: %bb.6(0x80000000), %bb.4(0x00000800); %bb.6(100.00%), %bb.4(0.00%)


bb.4 (machine-block-address-taken, landing-pad, ehfunclet-entry):
; predecessors: %bb.3
  successors: %bb.5(0x80000000); %bb.5(100.00%)

  CLEANUPRET

bb.5 (landing-pad, ehfunclet-entry):
; predecessors: %bb.0, %bb.4

  CLEANUPRET

bb.6:
; predecessors: %bb.3, %bb.0

  RET 0
```
After:
```
bb.0:
  successors: %bb.6(0x7ffff800), %bb.5(0x00000800); %bb.6(100.00%), %bb.5(0.00%)

  JMP_1 %bb.6

bb.3:
  successors: %bb.6(0x80000000), %bb.4(0x00000800); %bb.6(100.00%), %bb.4(0.00%)


bb.4 (machine-block-address-taken, landing-pad, ehfunclet-entry):
; predecessors: %bb.3, %bb.6
  successors: %bb.5(0x80000000); %bb.5(100.00%)

  CLEANUPRET

bb.5 (landing-pad, ehfunclet-entry):
; predecessors: %bb.0, %bb.4

  CLEANUPRET

bb.6:
; predecessors: %bb.3, %bb.0
  successors: %bb.4(0x00000800); %bb.4(0.00%)

  RET 0
```

phoebewang replied:
I see we assert for `(*SI)->isEHPad()`, but `bb.6` is not a EHPad. Did I misunderstand something here?
BTW, can we use `SI->isEHPad()` directly?

HaohaiWen replied:
bb.6 is CurTBB so it will be skipped.
using pred_iterator = std::vector< MachineBasicBlock * >::iterator
Dereference pred_iterator get MachineBasicBlock *.

phoebewang reviewed (approved):
LGTM.

dyung commented on llvm/test/CodeGen/X86/windows-seh-EHa-PreserveCFG.ll:
```diff
@@ -0,0 +1,80 @@
+; RUN: llc -mtriple=x86_64-pc-windows-msvc %s
```
This RUN line seems to leave a .s file around that causes failures in subsequent runs:
https://lab.llvm.org/buildbot/#/builders/139/builds/56922
```
******************** TEST 'LLVM :: CodeGen/X86/windows-seh-EHa-PreserveCFG.s' FAILED ********************
Test has no 'RUN:' line
********************
```
Is this actually testing anything? Or is it just trying to verify the compiler does not crash?

dyung replied:
@HaohaiWen Could you restructure this test so that it doesn't generate the .s file (maybe redirect the output to /dev/null or something similar?) This is causing unresolved test errors on several of the buildbots I manage.

HaohaiWen replied:
Oh, sorry. I fixed it in https://github.com/llvm/llvm-project/pull/77784.
It's verifying compiler not crash.

vvereschaka commented:
@HaohaiWen ,
looks like these changes break some builders with the failed test
`UNRESOLVED: LLVM::windows-seh-EHa-PreserveCFG.s`

https://lab.llvm.org/buildbot/#/builders/234
https://lab.llvm.org/buildbot/#/builders/58
https://lab.llvm.org/buildbot/#/builders/104
https://lab.llvm.org/buildbot/#/builders/235

>Oh, sorry. I fixed it in https://github.com/llvm/llvm-project/pull/77784.
it didn't fix the problem.

HaohaiWen commented:
> @HaohaiWen , looks like these changes break some builders with the failed test `UNRESOLVED: LLVM::windows-seh-EHa-PreserveCFG.s`
> 
> https://lab.llvm.org/buildbot/#/builders/234 https://lab.llvm.org/buildbot/#/builders/58 https://lab.llvm.org/buildbot/#/builders/104 https://lab.llvm.org/buildbot/#/builders/235
> 
> > Oh, sorry. I fixed it in #77784.
> > it didn't fix the problem.

I saw it. I've fixed this issue.
We may need to manually delete LLVM::windows-seh-EHa-PreserveCFG.s on those machines.
Do you know who can help us to delete it?

vvereschaka commented:
@HaohaiWen 
>We may need to manually delete LLVM::windows-seh-EHa-PreserveCFG.s on those machines.
ok, got it. I'll do it for those builders.

>Do you know who can help us to delete it?
if you see some other failed builders because of this problem you can find their owners on the `Workers` tab, `Admin` column
https://lab.llvm.org/buildbot/#/workers

vvereschaka commented:
also, you can try to start a `Force Build` with cleaning of the source/build directories for these builders. It should clean up the current source folder and reload it.

HaohaiWen commented:
> also, you can try to start a `Force Build` with cleaning of the source/build directories for these builders. It should clean up the current source folder and reload it.

Thanks! I saw it has been forced to do clean up build.
